sup3r/models/abstract.py

```diff
@@ -424,3 +657,290 @@ def early_stop(history, column, threshold=0.005, n_epoch=5):
+        epochs : list
+            List of epochs being iterated through
+        t0 : float
+            Starting time of training.
+        loss_details : dict
+            Namespace of the breakdown of loss components
+        checkpoint_int : int | None
+            Epoch interval at which to save checkpoint models.
+        out_dir : str
+            Directory to save checkpoint models. Should have {epoch} in
+            the directory name. This directory will be created if it does not
+            already exist.
+        early_stop_on : str | None
+            If not None, this should be a column in the training history to
+            evaluate for early stopping (e.g. validation_loss_gen,
+            validation_loss_disc). If this value in this history decreases by
+            an absolute fractional relative difference of less than 0.01 for
+            more than 5 epochs in a row, the training will stop early.
+        early_stop_threshold : float
+            The absolute relative fractional difference in validation loss
+            between subsequent epochs below which an early termination is
+            warranted. E.g. if val losses were 0.1 and 0.0998 the relative
+            diff would be calculated as 0.0002 / 0.1 = 0.002 which would be
+            less than the default thresold of 0.01 and would satisfy the
+            condition for early termination.
+        early_stop_n_epoch : int
+            The number of consecutive epochs that satisfy the threshold that
+            warrants an early stop.
+        extras : dict | None
+            Extra kwargs/parameters to save in the epoch history.
+
+        Returns
+        -------
+        stop : bool
+            Flag to early stop training.
+        """
+
+        self.log_loss_details(loss_details)
+
+        self._history.at[epoch, 'elapsed_time'] = time.time() - t0
+        for key, value in loss_details.items():
+            if key != 'n_obs':
+                self._history.at[epoch, key] = value
+
+        last_epoch = epoch == epochs[-1]
+        chp = checkpoint_int is not None and (epoch % checkpoint_int) == 0
+        if last_epoch or chp:
+            msg = ('Model output dir for checkpoint models should have '
+                   f'{"{epoch}"} but did not: {out_dir}')
+            assert '{epoch}' in out_dir, msg
+            self.save(out_dir.format(epoch=epoch))
+
+        stop = False
+        if early_stop_on is not None and early_stop_on in self._history:
+            stop = self.early_stop(self._history, early_stop_on,
+                                   threshold=early_stop_threshold,
+                                   n_epoch=early_stop_n_epoch)
+            if stop:
+                self.save(out_dir.format(epoch=epoch))
+
+        if extras is not None:
+            for k, v in extras.items():
+                self._history.at[epoch, k] = v
+
+        return stop
+
+
+# pylint: disable=E1101,W0201,E0203
+class AbstractWindInterface(ABC):
```

Comment:
Can't we have this subclass AbstractInterface and then override the methods instead of appending the _wind suffix? @grantbuster

Reply:
Related to `_tf_generate()` vs. `_tf_generate_wind()`? You could move `_tf_generate()` to the `AbstractSingleModel` and have `AbstractWindInterface(AbstractSingleModel)` but then you're trading an inheritance conflict for the diamond problem... I think i'd prefer the inheritance conflict handled explicitly rather than the diamond problem.

Reply:
The methods could also just be moved into WindCondMom instead of AbstractWindInterface

Reply:
I don't think so, the point of the `AbstractWindInterface` is to have a common mixin class for the wind models - `WindGan` and `WindCondMom`. If you put the methods in `WindCondMom` you wouldn't have them in `WindGan`

Reply:
Oh yeah I missed them in WindGan. Ok this is fine then!

Reply:
Yeah it's a messy web haha. I think this is the best of the not-great options.